plugins/inputs/statsd/datadog.go

```diff
@@ -0,0 +1,173 @@
+		return fmt.Errorf("Invalid event message format: empty 'title' or 'text' field")
+	}
+
+	// Handle hostname, with a priority to the h: field, then the host:
+	// tag and finally the defaultHostname value
+	// Metadata
+	m := cachedEvent{
+		name: rawTitle,
+	}
+	m.tags = make(map[string]string, strings.Count(message, ",")+2) // allocate for the approximate number of tags
+	m.fields = make(map[string]interface{}, 9)
+	m.fields["alert-type"] = "info" // default event type
+	m.fields["text"] = uncommenter.Replace(string(rawText))
+	m.tags["source"] = defaultHostname
+	m.fields["priority"] = priorityNormal
+	m.ts = now
+	if len(message) == 0 {
+		s.events = append(s.events, m)
+		return nil
+	}
+
+	if len(message) > 1 {
+		rawMetadataFields := strings.Split(message[1:], "|")
+		for i := range rawMetadataFields {
+			if len(rawMetadataFields[i]) < 2 {
+				log.Printf("W! [inputs.statsd] too short metadata field")
+			}
+			switch rawMetadataFields[i][:2] {
+			case "d:":
+				ts, err := strconv.ParseInt(rawMetadataFields[i][2:], 10, 64)
+				if err != nil {
+					log.Printf("W! [inputs.statsd] skipping timestamp: %s", err)
+					continue
+				}
+				m.fields["ts"] = ts
+			case "p:":
+				switch rawMetadataFields[i][2:] {
+				case priorityLow:
+					m.fields["priority"] = priorityLow
+				case priorityNormal: // we already used this as a default
+				default:
+					log.Printf("W! [inputs.statsd] skipping priority")
+					continue
+				}
+			case "h:":
+				m.tags["source"] = rawMetadataFields[i][2:]
+			case "t:":
+				switch rawMetadataFields[i][2:] {
+				case "error":
+					m.fields["alert-type"] = "error"
+				case "warning":
+					m.fields["alert-type"] = "warning"
+				case "success":
+					m.fields["alert-type"] = "success"
+				case "info": // already set for info
+				default:
+					log.Printf("W! [inputs.statsd] skipping alert type")
+					continue
+				}
+			case "k:":
+				// TODO(docmerlin): does this make sense?
+				m.tags["aggregation-key"] = rawMetadataFields[i][2:]
+			case "s:":
+				m.fields["source-type-name"] = rawMetadataFields[i][2:]
+			default:
+				if rawMetadataFields[i][0] == '#' {
+					parseDataDogTags(m.tags, rawMetadataFields[i][1:])
+				} else {
+					log.Printf("W! [inputs.statsd] unknown metadata type: '%s'", rawMetadataFields[i])
+				}
+			}
+		}
+	}
+	// host is a magic tag in the system, and it expects it to replace the result of h: if it is present
+	// telegraf will add a"host" tag anyway with different meaning than dogstatsd, so we need to switch these out
+	if host, ok := m.tags["host"]; ok {
+		delete(m.tags, "host")
+		m.tags["source"] = host
+	}
+	s.Lock()
+	s.events = append(s.events, m)
+	s.Unlock()
+	return nil
+}
+
+func parseDataDogTags(tags map[string]string, message string) {
+	start, i := 0, 0
+	var k string
+	var inVal bool // check if we are parsing the value part of the tag
+	for i = range message {
+		if message[i] == ',' {
+			if k == "" {
+				k = message[start:i]
+				tags[k] = ""
```

Comment:
I've done empty string tags before in the zipkin plugin, however, it makes it pretty difficult to query for at least with influxdb.

@danielnelson how are label type tags typically done in telegraf?

Reply:
What is the value of having an empty tag?

Reply:
I don't know but Datadog events tests in their agent specifically makes allowances for them.  We wouldn't have spec compatibility if we didn't allow them.  Should we handle them some other way?

Reply:
I don't think it is allowed to have an empty tag value in InfluxDB, we just toss them at serialize time.